File: python3/unique_morse_code_words.py
"""
Space   : O(n)
Time    : O(nm)
"""


class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        code = [".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--",
                "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."]

        ans = set()
        for word in words:
            morse = ''
            for i in word:
                morse += code[ord(i)-97]
            ans.add(morse)

        return len(ans)

File: golang/unique_morse_code_words.go
package golang

func uniqueMorseRepresentations(words []string) int {
    truth := []string{".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."}
    
    hashmap := make(map[string]int)
    n := len(words)
    
    for i := 0; i < n; i++ {
        var morse string
        for j := 0; j < len(words[i]); j++ {
            morse += truth[words[i][j]-97]
        }
        if _, ok := hashmap[morse]; ok {
            hashmap[morse]++
        } else {
            hashmap[morse] = 1
        }
    }
    
    return len(hashmap)
}